### PR TITLE
Some bindings for RuntimeEffect

### LINKED ISSUE #120: Expose SkRuntimeEffect API
Has there been any consideration for adding the SkRuntimeEffect API to write custom SkShaders? Are there any blockers?

### PR DESCRIPTION
**Pull request by Vechro**
I tried to add some basic bindings for RuntimeEffect but I think I need some advice on how to proceed.

1. How should I handle the Result struct in RuntimeEffect.h? Should I also be returning one in Java?
2. That MakeShader binding is a mess, I have no idea how to deal with `childrenPtrs` and `uniformPtr`.
3. Are there any more RuntimeEffect functions that need bindings?

I'd appreciate any feedback, C++ isn't my strong suit.

Reference material:
* https://skia.org/docs/user/sksl/
* https://github.com/mono/SkiaSharp/pull/1221/files#diff-9b53ac31984a05ad45ba4e2d111d9dc89289422983199bc68a2512ee750da04cR2896
* https://skia.googlesource.com/skia/+/refs/heads/main/modules/canvaskit/canvaskit_bindings.cpp#1727

Closes #120.

### CODE REVIEW

**Comment by tonsky** [comment]
Hi, thanks for the contribution!

I’ll try to familiriaze myself with RuntimeEffect next week and see if I can help

File: platform/cc/RuntimeEffect.cc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include "SkRuntimeEffect.h"
+#include "interop.hh"
+
+JNIEXPORT jlong JNICALL
+Java_org_jetbrains_skija_RuntimeEffect__1nMakeShader(JNIEnv* env,
+                                                     jclass jclass,
+                                                     jlong ptr,
+                                                     jlong uniformPtr,
+                                                     jlongArray childrenPtrs,
+                                                     jlong childCount,
```
Comment (tonsky) [inline platform/cc/RuntimeEffect.cc:12]: No need to pass count explicitly, java arrays already have count

File: platform/cc/RuntimeEffect.cc [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                                                     jfloatArray localMatrixArr,
+                                                     jboolean isOpaque) {
+    SkRuntimeEffect* runtimeEffect = jlongToPtr<SkRuntimeEffect*>(ptr);
+    SkData* uniform = jlongToPtr<SkData*>(uniformPtr);
+    sk_sp<SkData> uniformData = SkData::MakeFromMalloc(uniform, uniform->size());
+
+    // Matrix
+    jfloat* m = env->GetFloatArrayElements(localMatrixArr, 0);
```
Comment (tonsky) [inline platform/cc/RuntimeEffect.cc:20]: Consider using skija::skMatrix, it does what you do here + handles null pointer in the place of an array

File: platform/cc/RuntimeEffect.cc [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    SkMatrix* mPtr = new SkMatrix();
+    mPtr->setAll(m[0], m[1], m[2], m[3], m[4], m[5], m[6], m[7], m[8]);
+    env->ReleaseFloatArrayElements(localMatrixArr, m, 0);
+
+    sk_sp<SkShader>* children = new sk_sp<SkShader>[childCount];
```
Comment (tonsky) [inline platform/cc/RuntimeEffect.cc:25]: This array will leak. Consider using std::vector instead

File: platform/cc/RuntimeEffect.cc [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    SkShader** cPtrs = reinterpret_cast<SkShader**>(childrenPtrs);
```
Comment (tonsky) [inline platform/cc/RuntimeEffect.cc:26]: This won’t fly. Java arrays are not C arrays, unfortunately. You need to request special access to them. You need env->GetLongArrayElements, paired with ReleaseLongArrayElements after the loop. Each individual long can then be cast to pointer using jlongToPtr

File: platform/cc/RuntimeEffect.cc [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    for (size_t i = 0; i < childCount; i++) {
+        // This bare pointer was already part of an sk_sp (owned outside of here),
+        // so we want to ref the new sk_sp so makeShader doesn't clean it up.
+        children[i] = sk_ref_sp<SkShader>(cPtrs[i]);
+    }
+    sk_sp<SkShader> shader =
+        runtimeEffect->makeShader(uniformData, children, childCount, mPtr, isOpaque);
+    return ptrToJlong(&shader);
```
Comment (tonsky) [inline platform/cc/RuntimeEffect.cc:34]: should be `ptrToJlong(shader.release())`

File: platform/cc/RuntimeEffect.cc [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+JNIEXPORT jlong JNICALL Java_org_jetbrains_skija_RuntimeEffect__1nMakeColorFilter(JNIEnv* env,
+                                                                                  jclass jclass,
+                                                                                  jstring sksl) {
+    SkString skslProper = skString(env, sksl);
+    SkRuntimeEffect::Result result = SkRuntimeEffect::MakeForColorFilter(skslProper);
```
Comment (tonsky) [inline platform/cc/RuntimeEffect.cc:41]: Check for error and throw java exception in case of a failure

File: shared/java/RuntimeEffect.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+public class RuntimeEffect extends RefCnt {
+    static {
+        Library.staticLoad();
+    }
+
+    public Shader makeShader(@Nullable Data uniforms, @Nullable Shader[] children, @Nullable Matrix33 localMatrix,
+            boolean isOpaque) {
+        Stats.onNativeCall();
+        float[] arr = localMatrix == null ? null : localMatrix._mat;
+        long childCount = children == null ? 0 : children.length;
```
Comment (tonsky) [inline shared/java/RuntimeEffect.java:16]: make it int, will save you a cast on the next line

**Review by tonsky — CHANGES_REQUESTED** [review]
Looks good, great start! I have a few notes

File: platform/cc/RuntimeEffect.cc (hunk 4 above)
Comment (tonsky) [reply platform/cc/RuntimeEffect.cc:26]: Here’s an example of a similar situation https://github.com/JetBrains/skija/blob/8b2111460148cf6f190e5a1db24cf8ebfb069dcd/platform/cc/ImageFilter.cc#L126-L139

**Comment by Vechro** [comment]
I addressed all your comments as well as I could. I figure about now I should actually test if any of this works.

**Comment by tonsky** [comment]
Write a scene in the examples/scenes directory, add it to Scenes.java, then test with any of examples/lwjgl, examples/kwinit or examples/jwm (they all use the same scenes). It’s not automatic (check with your eyes), but works ok

**Comment by Vechro** [comment]
Okay, it's not going as smoothly as I anticipated. I may need your help figuring out the culprit here. I'm trying to recreate a simpler example of one of the examples on Skia's website, the SkSL should be correct as it works in the Skia sandbox. Yet I manage to get a `siginfo: EXCEPTION_ACCESS_VIOLATION (0xc0000005), reading address 0x0000000000000020` when I try to recreate it as a Scene.
I've included the full error log: [hs_err_pid25196.log](https://github.com/JetBrains/skija/files/6949384/hs_err_pid25196.log). The branch now also includes the example I wrote. Any insight would be greatly appreciated.

**Comment by tonsky** [comment]
Merged. Thanks for the contribution, from the words of Skia team at Google, I know it’s an important part of the Skia API, and finally Skija has it as well.

Also, check out the example!

**Comment by Vechro** [comment]
Thank you! The example looks great!

**Pull request merged**